core/src/jmh/java/com/uber/m3/tally/AbstractReporterBenchmark.java

```diff
@@ -0,0 +1,140 @@
+
+package com.uber.m3.tally;
+
+import com.uber.m3.util.Duration;
+import com.uber.m3.util.ImmutableMap;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+
+// TODO: 31/01/2021 add descrition and readme how to use.
+@State(Scope.Benchmark)
+public abstract class AbstractReporterBenchmark {
+    private static final ImmutableMap<String, String> DEFAULT_TAGS = new ImmutableMap.Builder<String, String>(5)
+            .put("tag1", "test1")
+            .put("tag2", "test2")
+            .put("tag3", "test3")
+            .put("tag4", "test4")
+            .put("env", "test")
+            .build();
+
+    private static final DurationBuckets DURATION_EXPONENTIAL_BUCKETS = DurationBuckets.linear(
+            Duration.ofMillis(1), Duration.ofMillis(10), 128
+    );
+
+    private static final ValueBuckets VALUE_EXPONENTIAL_BUCKETS = ValueBuckets.linear(
+            0.1, 100d, 128
+    );
+
+    private static final String COUNTER_NAME = "counter";
+    private static final String GAUGE_NAME = "gauge";
+    private static final String TIMER_NAME = "timer";
+    private static final String HISTOGRAM_DURATION_NAME = "histogram_duration";
+    private static final String HISTOGRAM_VALUE_NAME = "histogram_value";
+
+    private StatsReporter reporter;
+
+    @Benchmark
+    public void reportCounterBenchmark(IncrementalValueState value) {
+        reporter.reportCounter(COUNTER_NAME, DEFAULT_TAGS, value.incrementAndGet());
+    }
+
+    @Benchmark
+    public void reportGaugeBenchmark(IncrementalValueState value) {
+        reporter.reportCounter(COUNTER_NAME, DEFAULT_TAGS, value.incrementAndGet());
+    }
+
+    @Benchmark
+    public void reportTimerBenchmark(IncrementalValueState value) {
+        reporter.reportTimer(TIMER_NAME, DEFAULT_TAGS, Duration.ofSeconds(value.incrementAndGet()));
+    }
+
+    @Benchmark
+    public void reportHistogramDurationSamplesBenchmark(IncrementalValueState value) {
+        reporter.reportHistogramDurationSamples(
+                HISTOGRAM_DURATION_NAME,
+                DEFAULT_TAGS,
+                DURATION_EXPONENTIAL_BUCKETS,
+                DURATION_EXPONENTIAL_BUCKETS.getDurationLowerBoundFor(10),
+                DURATION_EXPONENTIAL_BUCKETS.getDurationLowerBoundFor(10),
+                value.incrementAndGet()
+        );
+    }
+
+    @Benchmark
+    public void reportHistogramValueSamplesBenchmark(IncrementalValueState value) {
+        reporter.reportHistogramValueSamples(
+                HISTOGRAM_VALUE_NAME,
+                DEFAULT_TAGS,
+                VALUE_EXPONENTIAL_BUCKETS,
+                5.0,
+                5.0,
+                value.incrementAndGet()
+        );
+    }
+
+    @Setup
+    public void setup() {
+        this.reporter = initReporter();
+
+        reporter.reportCounter(COUNTER_NAME, DEFAULT_TAGS, 1);
```

Comment:
Just to clarify--initial reporting is to make sure that we're not measuring the new metric setup, right?

Think it would be worth also benchmarking addition of new metrics? It's also something that can occur in the hotpath (though of course the expectations can be a bit less stringent perhaps than the repeat metric case).

Reply:
> Just to clarify--initial reporting is to make sure that we're not measuring the new metric setup, right?
yep, that's correct.
> Think it would be worth also benchmarking addition of new metrics? It's also something that can occur in the hotpath (though of course the expectations can be a bit less stringent perhaps than the repeat metric case).
I think it's a really good test in terms of memory usage, but not sure how much value we can get from throughput or overage time benchmarking. Also I think in this case average time is more useful then throughput

Reply:
> I think it's a really good test in terms of memory usage

Agreed yeah--that's what I'd want to see I think.